### PR TITLE
Move widget functions into matplotlib.testing.widgets.

### LINKED ISSUE #7372: Moving get_ax and do_event to testing
I'm wanting to do some testing in a library that uses matplotlib. In particular, this library is using matplotlib native widgets.

There appear to be some [utility functions]( https://github.com/matplotlib/matplotlib/blob/6e24047bdcdc0f35d282c27d62324e52fa7a7afb/lib/matplotlib/tests/test_widgets.py#L17-L76 ) used for testing called `get_ax` and `do_event`. These seem like they would be very handy for widget tests like the ones proposed above.

However, I'm not entirely clear on how they can access them ATM as they don't seem to be importable and they are not in `testing` currently. Would it be possible to move them to `testing` so they could be used externally? Related would that actually make sense/be usable for my use case?

### PR DESCRIPTION
**Pull request by SidharthBansal**
## PR Summary

Fixes #7372, move widget functions into matplotlib.testing.widgets.

Co-authored-by: @dakotablair
Closes #9905

## PR Checklist

- [ ] Has Pytest style unit tests
- [ ] Code is [Flake 8](http://flake8.pycqa.org/en/latest/) compliant
- [ ] New features are documented, with examples if plot related
- [ ] Documentation is sphinx and numpydoc compliant
- [ ] Added an entry to doc/users/next_whats_new/ if major new feature (follow instructions in README.rst there)
- [ ] Documented in doc/api/api_changes.rst if API changed in a backward-incompatible way

### CODE REVIEW

**Comment by SidharthBansal** [comment]
Travis fails with 
> E           matplotlib.testing.exceptions.ImageComparisonFailure: images not close (RMS 6.457):
E           	result_images/test_bbox_tight/bbox_inches_tight_raster_pdf.png
E           	result_images/test_bbox_tight/bbox_inches_tight_raster-expected_pdf.png

Same error in another #17028 
Is there any need to change bbox_inches_tight_raster-expected_pdf.png (baseline image)? I think that my changes are not related to this error. Can you guys kindly help? @anntzer @timhoffm

**Comment by anntzer** [comment]
I think that test has been failing for most of the day, we just need to investigate it separately...

**Comment by SidharthBansal** [comment]
Let's open up a separate issue and tackle it there.


Travis behave very wierdly many times.

**Comment by SidharthBansal** [comment]
@timhoffm kindly review and suggest changes if any.
Travis breaks due #17037

**Comment by tacaswell** [comment]
We have a preference for rebasing onto master rather than merging master in to the feature branch.

Could you rebase (and force push) this @SidharthBansal ?  The other option is we can squash merge.

**Comment by SidharthBansal** [comment]
Ok. I will squash it. Thanks

**Comment by SidharthBansal** [comment]
I rebased and forced push. Thanks!

File: lib/matplotlib/testing/widgets.py [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+See also matplotlib.tests.test_widgets
+"""
+
+import matplotlib.pyplot as plt
+
+try:
+    # mock in python 3.3+
+    from unittest import mock
+except ImportError:
+    import mock
```
Comment (QuLogic) [inline lib/matplotlib/testing/widgets.py:15]: We don't support Python 2 any more; you don't need this import.

File: lib/matplotlib/testing/widgets.py [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+
+
+def get_ax():
```
Comment (QuLogic) [inline lib/matplotlib/testing/widgets.py:18]: This needs a docstring.

File: lib/matplotlib/testing/widgets.py [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+    fig, ax = plt.subplots(1, 1)
+    ax.plot([0, 200], [0, 200])
+    ax.set_aspect(1.0)
+    ax.figure.canvas.draw()
+    return ax
+
+
+def do_event(tool, etype, button=1, xdata=0, ydata=0, key=None, step=1):
+    """
```
Comment (QuLogic) [inline lib/matplotlib/testing/widgets.py:27]: Can you convert this to numpydoc?

File: lib/matplotlib/testing/widgets.py [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+    Trigger an event
+
+    Parameters
+    ----------
+    tool : a matplotlib.widgets.RectangleSelector instance
```
Comment (QuLogic) [inline lib/matplotlib/testing/widgets.py:27]: The type is the type, not anything more.
```suggestion
    tool : matplotlib.widgets.RectangleSelector
```

File: lib/matplotlib/testing/widgets.py [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+        the key depressed when the mouse event triggered (see
+        :class:`KeyEvent`)
+    step : int
+        number of scroll steps (positive for 'up', negative for 'down')
+    """
+    event = mock.Mock()
+    event.button = button
+    ax = tool.ax
+    event.x, event.y = ax.transData.transform([(xdata, ydata),
+                                               (xdata, ydata)])[00]
```
Comment (QuLogic) [inline lib/matplotlib/testing/widgets.py:47]: Dunno why it was like this before.
```suggestion
                                               (xdata, ydata)])[0]
```

**Comment by QuLogic** [comment]
It's not really a breaking API change, as this is all new API.

**Comment by SidharthBansal** [comment]
Thanks @QuLogic. I have completed the suggested changes. :)

**Comment by tacaswell** [comment]
:sheep: I miss read the UI and thought @QuLogic had previously approved it, but that was me so I merged this with one review, sorry.

**Pull request merged**